clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/KafkaStoreIngestionService.java

```diff
@@ -1139,6 +1139,11 @@ public MetadataResponse getMetadata(String storeName) {
       response.setLatestSuperSetValueSchemaId(latestSuperSetValueSchemaId);
       response.setRoutingInfo(routingInfo);
       response.setHelixGroupInfo(helixGroupInfo);
+
+      // return response with all fields set to null if client has up-to-date metadata
+      if (response.hashCode() == hash) {
+        return new MetadataResponse();
+      }
```

Comment:
Hmm... by the time we get here, a ton of work has already gone into preparing the response. Are we gaining much by throwing away the response that's already there and returning this empty one instead? I guess the response we serialize is smaller this way...?

I wonder if we should keep a cache of `storeName -> MetadataResponse` that's already fully prepared (maybe even already serialized), and its hash pre-computed. This way none of the work here happens on the query path. All that would happen in the query path is one map lookup, comparing the hash, and based on that either sending out the empty response or the real response. The trick would be to have the right listeners in place to refresh this pre-computed state whenever any of the inputs change.

Anyway, all of that is likely a premature optimization at this point, though it could become important when we move to the header-based approach and this gets evaluated on ever data query...

Reply:
That's a good point, definitely need some logic in the server that's maintaining this cache of metadata on every metadata change event from ZK